Refactor variance ratio, cyclic_shift and confint.cyclic_shift

aammd opened the pull request:
Hello @laurenmh @mbjones et al,

I was attempting to respond to out reviewer feedback, particularly for #66 . I started puling on one thread, and soon I found myself making many changes. I was just about to update the tests but I thought it would be best to see what you think, first.

Major changes in this PR include:
* `variance_ratio` now does all the logic around responding to the value of `replicate.var` and `average.replicates`.
* `cyclic_shift` is now only for applying a given method to a single replicate, and returns the null vector in an object of class `cyclic_shift`
* `confint.cyclic_shift` calculates the two-tailed quantiles of that output.

Minor changes: 
* code formatting
* adds dependency on `purrr` for convenient ways to stitch lists together into `data.frames()`

Please let me know what you think! I'll add some questions as line notes

aammd commented on R/community_stability.R:
```diff
@@ -74,8 +74,15 @@ community_stability <- function(df, time.var="year", abundance.var="abundance",
+  assertthat::assert_that(assertthat::has_name(df, x))
+  assertthat::assert_that(is.numeric(df[[x]]))
+
+  sync_var <- df[[x]]
```
adds a few assertions and removes the wholesale import of `stats`

aammd commented on R/rank_shift.R:
```diff
@@ -76,32 +76,38 @@ mean_rank_shift <- function(df, time.var = "year", species.var = "species",
 
-    commonspp <- Map(df_intersect, y1, y2, dataname = species.var)
+  for (v in c(time.var, species.var, abundance.var)){
+    assertthat::assert_that(assertthat::has_name(comm_data, v))
```
assert that those names are actually present in the data (very weird errors if they are not!)

aammd commented on R/temporal_torus_translation.R:
```diff
@@ -14,16 +14,41 @@
+                         bootnumber){
+
+  ## match arg to check method name
+  ## switch to go from character to function
```
I was thinking of allowing people to specify their method using characters, rather than passing in the actual function . but perhaps unnecessary ? isn't `method` always going to be `variance_ratio_longdata` or `variance_ratio_matrixdata` ? do we even *want* users to use this function anymore?

laurenmh replied:
Hi Andrew - I think we want it to be an actual function - with the
restructure this will be a pretty general function, people might want to
use it. When its used in the variance_ratio it'll always be
variance_ratio_matrixdata

On Fri, Mar 18, 2016 at 12:59 AM, Andrew MacDonald <notifications@github.com
> wrote:

> In R/temporal_torus_translation.R
> <https://github.com/laurenmh/codyn/pull/67#discussion_r56619026>:
>
> >  #' @references
> >  #' Hallett, Lauren M., Joanna S. Hsu, Elsa E. Cleland, Scott L. Collins, Timothy L. Dickson, Emily C. Farrer, Laureano A. Gherardi, et al. "Biotic Mechanisms of Community Stability Shift along a Precipitation Gradient." Ecology 95, no. 6 (2014): 1693-1700.
> >  #'
> >  #' Harms, Kyle E., Richard Condit, Stephen P. Hubbell, and Robin B. Foster. "Habitat Associations of Trees and Shrubs in a 50-Ha Neotropical Forest Plot." Journal of Ecology 89, no. 6 (2001): 947-59.
> >  #' @export
> > -cyclic_shift <- function(df, time.var="year", species.var="species",  abundance.var="abundance", FUN, bootnumber){
> > -  if(!is.numeric(df[[abundance.var]])) { stop("Abundance variable is not numeric") }
> > -  out<-replicate(bootnumber, FUN(shuffle_community(transpose_community(df, time.var,  species.var, abundance.var))))
> > -  return(out)
> > +cyclic_shift <- function(df, time.var="year",
> > +                         species.var="species",
> > +                         abundance.var="abundance",
> > +                         method,
> > +                         bootnumber){
> > +
> > +  ## match arg to check method name
> > +  ## switch to go from character to function
>
> I was thinking of allowing people to specify their method using
> characters, rather than passing in the actual function . but perhaps
> unnecessary ? isn't method always going to be variance_ratio_longdata or
> variance_ratio_matrixdata ? do we even *want* users to use this function
> anymore?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/laurenmh/codyn/pull/67/files/f3e38639882cb3e07b489b62de4bd6214853b215#r56619026>
>

aammd replied:
hi @laurenmh ! sounds good. What are the possible values for `method` (formerly `FUN`)? I have seen `variance_ratio_matrixdata` and `variance_ratio_longdata` and I think also `cov`. Are there  others?

aammd commented on R/temporal_torus_translation.R:
```diff
@@ -14,16 +14,41 @@
+
+
+  # add the method part in here
+  shift <- structure(list(out = out), class = "cyclic_shift")
```
S3 class

aammd commented on R/temporal_torus_translation.R:
```diff
@@ -62,45 +87,43 @@ cyclic_shift <- function(df, time.var="year", species.var="species",  abundance.
+  #     output <- cbind(reps, lowerCI, upperCI, nullmean)
+  #     names(output)[2:3] <- c("lowerCI", "upperCI")
+  #   }
+  # }
```
all this should probably be deleted

laurenmh replied:
Yes, since the confint does it

On Fri, Mar 18, 2016 at 1:00 AM, Andrew MacDonald <notifications@github.com>
wrote:

> In R/temporal_torus_translation.R
> <https://github.com/laurenmh/codyn/pull/67#discussion_r56619100>:
>
> > +  #     out <- do.call("rbind", lout)
> > +  #     out <- colMeans(out)
> > +  #     lowerCI <- quantile(out, li)
> > +  #     upperCI <- quantile(out, ui)
> > +  #     nullmean <- mean(out)
> > +  #     output <- as.data.frame(cbind(lowerCI, upperCI, nullmean))
> > +  #     row.names(output) <- NULL
> > +  #   } else {
> > +  #     lowerCI <- do.call("rbind", lapply(lout, quantile, li))
> > +  #     upperCI <- do.call("rbind", lapply(lout, quantile, ui))
> > +  #     nullmean <- do.call("rbind", lapply(lout, mean))
> > +  #     reps <- unique(df[replicate.var])
> > +  #     output <- cbind(reps, lowerCI, upperCI, nullmean)
> > +  #     names(output)[2:3] <- c("lowerCI", "upperCI")
> > +  #   }
> > +  # }
>
> all this should probably be deleted
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/laurenmh/codyn/pull/67/files/f3e38639882cb3e07b489b62de4bd6214853b215#r56619100>
>

aammd commented on R/variance_ratio.R:
```diff
@@ -33,100 +33,162 @@
-   if(is.na(replicate.var)) {
-        check_single_onerep(df, time.var, species.var)
-        VR <- variance_ratio_longformdata(df, time.var, species.var, abundance.var)
+  if  (missing(replicate.var)) {
```
in `variance_ratio` I decided to test for missing, rather than NA, column names. it seems to me that if there is no replicate column, people would rather not specify it than specify it as NA

laurenmh replied:
Okay. Don't think it matters too much as long as the default was NA. We'll
have to change this across all the functions for consistency

On Fri, Mar 18, 2016 at 1:01 AM, Andrew MacDonald <notifications@github.com>
wrote:

> In R/variance_ratio.R
> <https://github.com/laurenmh/codyn/pull/67#discussion_r56619165>:
>
> >    # if no replicates, calculate single variance ratio
> > -   if(is.na(replicate.var)) {
> > -        check_single_onerep(df, time.var, species.var)
> > -        VR <- variance_ratio_longformdata(df, time.var, species.var, abundance.var)
> > +  if  (missing(replicate.var)) {
>
> in variance_ratio I decided to test for missing, rather than NA, column
> names. it seems to me that if there is no replicate column, people would
> rather not specify it than specify it as NA
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/laurenmh/codyn/pull/67/files/f3e38639882cb3e07b489b62de4bd6214853b215#r56619165>
>

aammd replied:
Hm, didn't realize we did this elsewhere. I'll change this back and we can talk about making this change in a separate issue/PR later

aammd commented on R/variance_ratio.R:
```diff
@@ -33,100 +33,162 @@
+    check_single(df, time.var, species.var, replicate.var)
+    ## if you give a replicate, it must be a factor. This gives users responsibility for order.
+    assertthat::assert_that(is.factor(df[[replicate.var]]))
+
```
There used to be a line which checked `replicate.var` for being a factor, and if it was not (i think?) it coerced it. Here i'm suggesting we just force users to make it a factor themselves (safer, in case there is a particular order they want/expect in the output)

laurenmh replied:
I disagree - I'd rather that it could be either a character or factor, and
generally I think it should be a character. We only coerce it a factor if
it was a factor and there were factor levels that weren't represented in
the dataset. I did this because our big test dataset had this scenario a
lot, I'd hope that most datasets dont.

On Fri, Mar 18, 2016 at 1:04 AM, Andrew MacDonald <notifications@github.com>
wrote:

> In R/variance_ratio.R
> <https://github.com/laurenmh/codyn/pull/67#discussion_r56619352>:
>
> > +                            abundance.var = abundance.var,
> > +                            method = variance_ratio_matrixdata,
> > +                            bootnumber = bootnumber)
> > +
> > +    nullout <- confint(nullval)
> > +
> > +    # bind actual value and CI
> > +    output <- cbind(nullout, VR)
> > +
> > +  } else {
> > +
> > +    # if multiple replicates, check all replicates have values
> > +    check_single(df, time.var, species.var, replicate.var)
> > +    ## if you give a replicate, it must be a factor. This gives users responsibility for order.
> > +    assertthat::assert_that(is.factor(df[[replicate.var]]))
> > +
>
> There used to be a line which checked replicate.var for being a factor,
> and if it was not (i think?) it coerced it. Here i'm suggesting we just
> force users to make it a factor themselves (safer, in case there is a
> particular order they want/expect in the output)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/laurenmh/codyn/pull/67/files/f3e38639882cb3e07b489b62de4bd6214853b215#r56619352>
>

aammd replied:
OH I'm sorry, i misunderstood that! :flushed: I'll put it back the way I found it

aammd commented on R/variance_ratio.R:
```diff
@@ -33,100 +33,162 @@
+      VR_list <- lapply(X, FUN = variance_ratio_longformdata,
+                        time.var, species.var, abundance.var)
+
+      VRout <- purrr::map_df(VR_list, ~data.frame(VR = .), .id = replicate.var)
```
here are 2/3 of the places where I lean on `purrr::map_df`. I was running out of time and this was the best I could think of to solve this problem. I like that it let's us recover the name of `replicate.var`. What do you think? I know we are striving for minimal dependencies

laurenmh replied:
Oh cool I didn't know about that package! But since I don't know it, let's
check with Matt before building the dependency

On Fri, Mar 18, 2016 at 1:06 AM, Andrew MacDonald <notifications@github.com>
wrote:

> In R/variance_ratio.R
> <https://github.com/laurenmh/codyn/pull/67#discussion_r56619448>:
>
> > +      ## also do separate null models (ie on split list)
> > +      null_list <- lapply(X, FUN = cyclic_shift,  time.var = time.var,
> > +                          species.var = species.var,
> > +                          abundance.var = abundance.var,
> > +                          method = variance_ratio_matrixdata,
> > +                          bootnumber = bootnumber)
> > +      ## calculate confidence intervals for each:
> > +      null_intervals <- lapply(null_list, confint)
> > +
> > +      ## combine as data.frames and preserve replicate var
> > +      nullout <- purrr::map_df(null_intervals, as.data.frame, .id = replicate.var)
> > +
> > +      VR_list <- lapply(X, FUN = variance_ratio_longformdata,
> > +                        time.var, species.var, abundance.var)
> > +
> > +      VRout <- purrr::map_df(VR_list, ~data.frame(VR = .), .id = replicate.var)
>
> here are 2/3 of the places where I lean on purrr::map_df. I was running
> out of time and this was the best I could think of to solve this problem. I
> like that it let's us recover the name of replicate.var. What do you
> think? I know we are striving for minimal dependencies
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/laurenmh/codyn/pull/67/files/f3e38639882cb3e07b489b62de4bd6214853b215#r56619448>
>

mbjones replied:
@laurenmh @aammd purrr has dependencies on a number of other large packages, including magrittr (>= 1.5), dplyr (>= 0.4.3), Rcpp, lazyeval.  So this definitely increases the overhead of installing codyn.  That said, these are pretty standard packages these days, so probably not a big deal for most people.  Each time a package gets updated in some incompatible way, we will have to release a new version of codyn that fixes the issue. This happens more than you might think (most recently with Hadley's `testthat`. So just be prepared for more maintenance time with each dependency we add.

aammd replied:
@mbjones I'm disappointed that `purrr` depends on those big packages! I didn't realize that when I suggested it. Let's change it to something in base.. any suggestions? I liked that this method let me keep the name for `replicate.var`

aammd commented on R/variance_ratio.R:
```diff
@@ -33,100 +33,162 @@
+
+      for (B in 1:bootnumber) {
+        null_means[B] <- mean(vapply(null_output, `[[`, 1, i = B))
+      }
```
this for loop is probably needless. better to do `do.call(rbind, null_output)` followed by `colMeans`?

But in general, am I doing this correctly? right now we do this:
* split df based on replicate.var
* cyclic shift each replicate _independently_, obtaining one null distribution of variance ratio per replicate
* calculating the average variance ratio for all replicates -- all of the first numbers of the null distribution in each replicate, then all the second numbers, then all the third numbers etc
* average, quantiles of this vector of means.

Did I interpret that correctly? I am worried that it is incorrect. Shouldn't the same cyclic shift be used within each replicate that are combined in the same null mean? in other words, if we have four years of data (1, 2, 3, 4) and two replicates (A and B), shouldn't each randomization be the same in both?
A: 3,4,1,2
B: 3,4,1,2

but right now it could be something like
A: 3,4,1,2
B: 2,3,4,1

laurenmh replied:
Good point, but not all replicates will have the exact same species or same number of species, so you couldn't apply the same cyclic shift to them all. I think the latter scenario is fine.

gavinsimpson replied:
@laurenmh But shouldn't they though, have the same species and therefore the same number of species? At least that's how we do things in the multivariate ordination world with vegan etc? Sure there are issues with assuming that an unobserved species is a true 0 abundance or absence.

@aammd The argument goes that the same cyclic shift should be applied to the same set of sites/samples if you consider them to have experienced the same time. This massively reduces the number of potential permutations though, so people often ignore it. In your example there are only 4 permutations total, no matter how many replicates you have.

laurenmh replied:
@gavinsimpson @aammd Currently we cast a new species x time matrix for each replicate, and don't include species that are never present in the replicate. We did this partly to reduce data frame size, and partly to encourage tidy input data/model how to manipulate it. Some of the syntheses groups that we worked with to test the package had huge species x (site, time) matrices with most species in any given site being all 0s - this was frustrating and super hard to QA, so we opted to require the input data be formatted differently. 

Logistically it would require some overhaul, but we could shift to casting with all species in the overall dataset (thus, all replicates having the same species, that are sometimes all 0), then applying the same random starting points for each species across replicates. To do this I think we'd go back to appending the dataframe to itself, and saving/applying starting points - could you do this with shuffleSeries? Conceptually, though, I think it's pretty standard to have each replicate re-randomized to increase permutation number. You two okay with keeping it that way, and clarifying the help doc and MS text to spell out that that is what it is doing?

aammd commented:
OK here are a few things I'd like to do before merging this PR

- [ ] remove dependency on purrr but keep that behaviour
- [ ] put `replicate.var` back as NA (for now anyway)
- [ ] return the factor assertion in `variance_ratio` to its previous behaviour (ie allowing either factor or character)

Anything else? As I understand we have to finish this today -- is there a particular time we need to aim for?

laurenmh commented:
That sounds good - the original had the purr behavior (keeping the
replicate names) - this relied on us sorting the dataframe up top, but I
think okay.

Yes, for now i'd put replicate.var back, and keep replicate as either a
factor or character in variance_ratio

That's it! I got pretty sick last week, asked for an extension. Could you
get the PR ready today, and I'll do the rest of the cleaning tomorrow?

Thanks!

Lauren



On Fri, Mar 18, 2016 at 1:12 PM, Andrew MacDonald <notifications@github.com>
wrote:

> OK here are a few things I'd like to do before merging this PR
>
>    - remove dependency on purrr but keep that behaviour
>    - put replicate.var back as NA (for now anyway)
>    - return the factor assertion in variance_ratio to its previous
>    behaviour (ie allowing either factor or character)
>
> Anything else? As I understand we have to finish this today -- is there a
> particular time we need to aim for?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/laurenmh/codyn/pull/67#issuecomment-198458289>
>

aammd commented:
Hi @laurenmh , whew, relieved to hear about the extension (but sorry to hear you were sick! glad you are getting better).
Yes, I can take care of this today! :sweat_smile:

laurenmh commented:
Awesome, thanks Andrew!

On Mon, Mar 21, 2016 at 4:00 PM, Andrew MacDonald <notifications@github.com>
wrote:

> Hi @laurenmh <https://github.com/laurenmh> , whew, relieved to hear about
> the extension (but sorry to hear you were sick! glad you are getting
> better).
> Yes, I can take care of this today! [image: :sweat_smile:]
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/laurenmh/codyn/pull/67#issuecomment-199449914>
>

aammd commented:
Alright, I think this addresses most of the points of feedback! 

* removes dependency on `purrr`
* streamlines the `cyclic_shift`, applying suggestions from [Lauren's gist](https://gist.github.com/laurenmh/2cb97417b488731e5ead)
* adapts `variance_ratio` to handle the new behaviour of `cyclic_shift` and to take advantage of `confint`

I made a few minor adaptations to make it all fit together -- i'll try to highlight the ones that I'm most uncertain of.

aammd commented on R/variance_ratio.R:
```diff
@@ -33,100 +33,169 @@
+  } else {
+
+    ## drop any unusued levels
+    df <- droplevels(df)
```
I thought this might be a good way to drop unused factor levels (this is from base R). It seems to work when `replicate.var` is a character too

aammd commented on R/variance_ratio.R:
```diff
@@ -33,100 +33,169 @@
+      ## workaround, necessary because you are not allowed to pass an argument called FUN to lapply
+      cyclic_shift_nofun <- function(f = variance_ratio_matrixdata, ...){
+        function(...) {
+          cyclic_shift(FUN = f, ...)
```
This little bit of kludge is necessary because of `lapply()` Does it look OK?

aammd commented on R/variance_ratio.R:
```diff
@@ -33,100 +33,169 @@
+
+      ## use the replicate.var name
+      names(nullout)[1] <- replicate.var
+
```
These three lines of code are my replacement for `purrr::map_df`. I feel like it is a bit overcomplex, but couldn't come up with anything simpler :persevere: thoughts?

laurenmh replied:
works for me, prefer it to a large dependency, thanks!

aammd commented on R/utilities.R:
```diff
@@ -73,19 +73,21 @@ check_numeric <- function(df, time.var, abundance.var) {
-       please remove these replicates prior to calculations ")}
 
+  df <- droplevels(df)
+
```
Added droplevels here too, to ensure it always works correctly
